session/src/main/java/org/apache/iotdb/session/Session.java

```diff
@@ -585,6 +641,14 @@ private void handleRedirection(String deviceId, EndPoint endpoint)
     }
   }
 
+  private void handleQueryRedirection(EndPoint endPoint) throws IoTDBConnectionException {
+    if (enableQueryRedirection) {
+      defaultSessionConnection.close();
+      defaultSessionConnection = constructSessionConnection(this, endPoint, zoneId);
+      defaultSessionConnection.setEnableRedirect(enableQueryRedirection);
+    }
+  }
```

Comment:
Maybe it would be better to cache a certain number of connections for different endpoints.
If the client queries root.sg1, then root.sg2, then root.sg1, then root.sg2... a new connection will be created for each query if the two SGs are located in different data groups.

Reply:
yes, we could cache a certain number of connections for different node.